kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,348 @@
+-   `<digest>`: a lower-case hex-coded SHA-256 digest (JSON: string).
+-   `<file-path>`: a file path (JSON: string).
+-   `<file-path-list>`: a list of `file-path` (JSON: list).
+-   `<vname>`: a `kythe.proto.VName` message (JSON: object).
+-   `<protocol>`: a protocol version label (JSON: string).
+-   `<analysis-type-list>`: a list of an identifier for a specific type or types of analyses.
+-   `<output-encoding>`: an identifier indicating how the output will be encoded, must be one of `json` or `protobuf` (JSON: string).
+
+Where the description refers to a protocol buffer message, the message is
+encoded using the canonical
+[JSON mapping](https://developers.google.com/protocol-buffers/docs/proto3#json)
+for the named message type.
+
+#### analysis-type
+
+Analysis types are described by the following grammar:
+
+```grammatical-framework
+analysis-type         = analysis-type-segment 0*{"/" analysis-type-segment}
+                
+analysis-type-segment = unreserved -- RFC 3987
+```
+ 
+Analysis types whose leading segment is `kythe` are reserved for use by
+Kythe. An example analysis type might be `kythe/index/java`, indicating the
+Kythe indexer for Java.
+
+### Methods
+
+A driver defines the following JSON-RPC methods:
+
+-   `init(InitRequest) → InitReply`
+
+    The `init` method is a protocol handshake, and must be the first method
+    called by the analyzer upon startup. The request specifies the protocol
+    version understood by the analyzer. If the environment understands the
+    requested version, it echoes the version in its reply.
+
+    If any other method is called prior to a successful invocation of `init`, or
+    if the environment doesn't recognize the protocol version requested, the
+    environment will report ProtocolError and close the input channel.
+
+    This document defines the protocol `"kythe1"`.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":1,"method":"init","params":{"protocol":"kythe1", "output-encoding":"json"}}
+        « {"jsonrpc":"2.0","id":1,"result":{"protocol":"kythe1"}}
+
+-   `analyze(AnalyzeRequest) → AnalyzeReply`
+
+    Request a fresh compilation record from the environment, blocking until one
+    is available. The environment assigns a unique integer analysis ID to each
+    request, distinct from any other ID currently assigned to that analyzer.
+    Once an analysis ID has been assigned, it is designated as **pending** until
+    the analyzer sends a `done` notification (see below) for that ID. A driver
+    SHOULD NOT issue the same compilation record again while there is a pending
+    analysis ID for that record.
+    
+    The reply includes the input file(s) on which to begin analysis,
+    as well as the output file to which to emit analysis output. It
+    reflects some of the kythe.proto.CompilationUnit structure. The
+    driver (and driver environment) MUST make contents available for
+    each file referenced in the compilation record at the file system
```

Comment:
This is actually a little tricky. The compilation unit records _physical_ paths (in the `info` field), as they were spelled by the build system at the moment when the data were captured. Those paths often assume the existence of a file tree with symlinks pointing outside the repository root.

To actually set that up you'd need to solve a set of constraints: Your virtual (or actual) filesystem root has to be high enough that all the up-pointers have something to traverse, and that down-steps are correctly populated within that hierarchy. 

If the analyzer already virtualizes file access, then this is easy: You just map the path it requests directly to the data in the compilation record. But allowing the analyzer to access the filesystem through the system API means you have to actually create those structures (links, directories, etc.) and get their access permissions semi-right.

This isn't a dealbreaker, but it is an aspect of this design that requires much more detail that the old one didn't need.

Reply:
I want to make sure I understand your point here. If you're "simply" saying that the driver may need to virtualize an arbitrary portion of the filesystem, I agree. That's effectively what the in-analyzer file system virtualization is doing.

What I (naively) imagined is that we'd set up the virtualization mapping precisely for every path mentioned in the compilation record. I believe(d) that this is independent of symlinks, or permissions -- that the driver virtualization gets first crack at any path. 

If that's *not* the case, i.e. the driver is further downstream, then I agree there are some more subtleties here.

Reply:
I've spent a lot of time thinking about this, and I don't see the issue. But I'm new here, so clearly there's plenty of room for ignorance and misunderstanding ...

The compilation record stores, along with the byte content, the actual path (possibly relativized, and maybe bastardized for e.g. jar files) requested from the host OS, i.e. before any symlink/hardlink/overlay/blah is traversed/resolved. It's precisely that path that the virtualization layer will intercept requests for. Why is there anything to worry about here?

Reply:
Ah. What you're pointing out is if the idea is to leave the "original compiler code" essentially intact, then the analyzer may do all sorts of operations other than open(name, "r") on the files it actually requires. And, the logic follows that we attempt to circumvent these other operations, then we're still doing some amount of intra-analyzer virtualization.

I wonder how much this actually occurs in practice, though.

Reply:
> I wonder how much this actually occurs in practice, though.

I don't know. But I have seen some compilers do some odd things with the filesystem in order to obtain locking, memory-map files, and so on. Hopefully most of that stuff is ancillary to reading input files and can be ignored—but I do think this spec will need to spend some more time laying out the rules for what the virtualization promises (and thus what the analyzer can rely on). When we were going to make the analyzer figure out how to inject itself into the I/O mechanism of the underlying compiler explicitly, these things didn't need much discussion.

Reply:
Implementing any and all of the weird POSIX or non-POSIX-but-damn-the-torpedoes corner cases expected by compilers-used-to-implement-analyzers is unlikely to be feasible.  In the real world, I expect most compilers will fall into one of three categories:

1) Compilers which do weird corner-casey-crap but have a Filesystem abstraction which can be implemented by the analyzer to not do so (or use a custom protocol directly as would be required by RFC2966).
2) Compilers which do straightforward create/write open/read operations (but have no filesystem abstraction)
3) Lost causes

Both this RFC and RFC2966 can support 1 with roughly the same amount of work, but this or something similar is required for 2. Given this I'd be very tempted to piggyback on the operations supported by something like 9P2000 if only because there are a broad set of implementations on a lot of platforms and it has better specified (and more limited) semantics.  The "file operations" specified by RFC2966 are unlikely to be sufficient for pretending to be a filesystem.  That said, I suspect including an RPC for emitting analysis output rather than necessitating the using the filesystem for output (rather than only input) will be simpler and sufficient.

The driver will also have some limited ability to rewrite the paths it sends to the analyzer.  It *should* also have a pretty good idea of what files the analyzer may try to open as it shouldn't be opening anything not in the CU.

Reply:
I'll admit that the main reason I went with filesystem for output rather than RPC is pushback-on-multiple-encodings. I don't have a strong reason for pushing the filesystem as long as the output encoding can be negotiated (and efficiently supports proto delimited, in addition to JSON), except it keeps the rpc protocol simple and pure text.

Reply:
It is certainly true that using a for _output_ is a separate concern from using a virtual filesystem for _input_. I think the latter is the better-motivated case, and doesn't strictly need to be tied to the former.

Reply:
Either way though, my main point here was that the spec should probably say more about what the VFS will provide.